### ISSUE TITLE
Detect invalid overrides of TypedDict items

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#3097]
We need to implement various invalid-override checks for TypedDicts.

- [ ] [Overriding with a different item type](https://shark.fish/typing-conformance-report/typeddicts_inheritance.html#L48):

  ```py
  # > Changing a field type of a parent TypedDict class in a subclass is
  # > not allowed.
  
  class X1(TypedDict):
     x: str
  
  class Y1(X1):  # E[Y1]
  Expected a ty diagnostic for this line (tag 'Y1')
     x: int  # E[Y1]: cannot overwrite TypedDict field "x"
  ```
- [ ] [Merging two TypedDicts with inconsistent item types
](https://shark.fish/typing-conformance-report/typeddicts_inheritance.html#L58)

  ```py
  # > Multiple inheritance does not allow conflict types for the same name field:
  class X2(TypedDict):
     x: int
  
  class Y2(TypedDict):
     x: str
  
  class XYZ2(X2, Y2):  # E: cannot overwrite TypedDict field "x" while merging
     xyz: bool
  ```

   See also [this test](https://shark.fish/typing-conformance-report/typeddicts_readonly_inheritance.html#L119) and [this test](https://shark.fish/typing-conformance-report/typeddicts_readonly_inheritance.html#L132).
- [ ] [Overriding of `ReadOnly` items with subtypes is allowed, but not with inconsistent types](https://shark.fish/typing-conformance-report/typeddicts_readonly_inheritance.html#L50)
  ```py
  # > Subclasses can narrow value types of read-only items.
  
  
  class AlbumCollection(TypedDict):
      albums: ReadOnly[Collection[Album1]]
      alt: ReadOnly[list[str | int]]
  
  
  class RecordShop(AlbumCollection):
      name: str
      albums: ReadOnly[list[Album1]]  # OK
      alt: ReadOnly[list[str]]  # E
  ```
- [ ] [Overriding a `Required[int]` field with a `ReadOnly[int]` field should be disallowed](https://shark.fish/typing-conformance-report/typeddicts_readonly_inheritance.html#L94)
- [ ] [Overriding a `ReadOnly[Required[int]]` with a `ReadOnly[NotRequired[int]]` should be disallowed](https://shark.fish/typing-conformance-report/typeddicts_readonly_inheritance.html#L106)

TypedDicts are also briefly mentioned in https://github.com/astral-sh/ty/issues/2158, but I wanted to have a separate issue that we can label as conformance / prioritize independently.

**Comment by AlexWaygood** [comment]
I'm hopeful that the first and second bullet point here should just fall out naturally from a generalised Liskov implementation for non-methods that checks for overridden variable annotations. Those would be errors even if it were a normal class, not a TypedDict.

The third bullet will require TypedDict-specific special casing, however.

**Comment by sharkdp** [comment]
> I'm hopeful that the first and second bullet point here should just fall out naturally from a generalised Liskov implementation for non-methods that checks for overridden variable annotations. Those would be errors even if it were a normal class, not a TypedDict.

Right, see my last paragraph. I also made it ~~a child issue of~~ blocked by #2158.

**Comment by sharifhsn** [comment]
All of the specific examples mentioned above are closed by https://github.com/astral-sh/ruff/pull/24338. Despite some effort, I couldn't come up with an example of an invalid override of `TypedDict` that is not fixed by that implementation.
